/*
 * Copyright 2010-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

option java_package = "org.jetbrains.jet.lang.resolve.objc";
option java_generic_services = false;

message Function {
    required string name = 1;
    required string return_type = 2;
    repeated Parameter parameter = 3;

    message Parameter {
        required string name = 1;
        required string type = 2;
    }
}

message ObjCMethod {
    required Function function = 1;
    required bool class_method = 2;
}

message ObjCProperty {
    required string name = 1;
    required string type = 2;
    // TODO: attributes like copy/readonly/nonatomic
}

message ObjCClass {
    required string name = 1;
    optional string base_class = 2;
    repeated string protocol = 3;
    repeated string category = 4;
    repeated ObjCMethod method = 5;
    repeated ObjCProperty property = 6;
}

message ObjCProtocol {
    required string name = 1;
    repeated string base_protocol = 2;
    repeated ObjCMethod method = 3;
    repeated ObjCProperty property = 4;
}

message ObjCCategory {
    required string name = 1;
    repeated string base_protocol = 2;
    repeated ObjCMethod method = 3;
    repeated ObjCProperty property = 4;
}

message TranslationUnit {
    repeated ObjCClass class = 1;
    repeated ObjCProtocol protocol = 2;
    repeated ObjCCategory category = 3;
}
